File: src/main/year2020/Day01.java
package year2020;

import template.Day;

import java.util.List;

public class Day01 extends Day {


    public Day01() {
        super("input.txt", "01", "2020");
    }

    private List<Integer> getInputAsInt()  {
        List<String> lines = getInput();
        return lines.stream().map(Integer::parseInt).toList();
    }

    @Override
    public long part1() {
        List<Integer> input = getInputAsInt();
        for (int i = 0; i < input.size(); i++) {
            for (int j = i + 1; j < input.size(); j++) {
                if (input.get(i) + input.get(j) == 2020) {
                    return (long) input.get(i) * input.get(j);
                }
            }
        }
        throw new IllegalStateException("No two numbers found that add up to 2020");
    }

    public long part1V2() {
        List<Integer> input = getInputAsInt();
        for (int number : input) {
            int complement = 2020 - number;
            if (input.contains(complement)) {
                return (long) number * complement;
            }
        }
        throw new IllegalStateException("No two numbers found that add up to 2020");
    }

    @Override
    public long part2() {
        List<Integer> input = getInputAsInt();
        for (int i = 0; i < input.size(); i++) {
            for (int j = i + 1; j < input.size(); j++) {
                for (int k = j + 1; k < input.size(); k++) {
                    if (input.get(i) + input.get(j) + input.get(k) == 2020) {
                        return (long) input.get(i) * input.get(j) * input.get(k);
                    }
                }
            }
        }
        throw new IllegalStateException("No three numbers found that add up to 2020");
    }

}

File: src/main/year2020/Day01.kt
package main.year2020

import java.io.File

class Day01(text : String = "input.txt") {

    private val day = "01"
    private val year = "2020"

    private val input = File("src/resources/$year/$day/$text").readLines().map { it.toInt() }.toList()

    fun part1(): Int {
        for (i in input.indices) {
            for (j in i + 1 until input.size) {
                if (input[i] + input[j] == 2020) {
                    return input[i] * input[j]
                }
            }
        }
        throw IllegalStateException("No two numbers found that add up to 2020")
    }

    fun part1V2(): Int = input.filter { input.contains(2020 - it) }.let { it[0] * it[1] }

    fun part2(): Int {
        for (i in input.indices) {
            for (j in i + 1 until input.size) {
                for (k in j + 1 until input.size) {
                    if (input[i] + input[j] + input[k] == 2020) {
                        return input[i] * input[j] * input[k]
                    }
                }
            }
        }
        throw IllegalStateException("No three numbers found that add up to 2020")
    }

}

